pkg/networkservice/common/updatepath/common.go

```diff
@@ -17,74 +17,77 @@
+		2.3 if path has next segment available, but next name is not equal to segmentName, will return error.
+	3. if Index == 0, and there is no current segment will add one.
+*/
+func updatePath(conn *networkservice.Connection, segmentName string) (*networkservice.Connection, error) {
 	if conn == nil {
-		return errors.New("NetworkServiceRequest.Connection must not be nil")
+		conn = &networkservice.Connection{}
 	}
 
 	// If we don't have a Path, add one
-	if conn.GetPath() == nil {
-		conn.Path = &networkservice.Path{}
+	if conn.GetPath() == nil || len(conn.GetPath().GetPathSegments()) == 0 {
+		if conn.Path == nil {
+			conn.Path = &networkservice.Path{}
+		}
+		conn.Path.Index = 0
+		// Replace path for first segment.
+		conn.Path.PathSegments = []*networkservice.PathSegment{
+			{
+				Name: segmentName,
+				Id:   conn.Id,
+			},
+		}
+		return conn, nil
 	}
 
 	path := conn.GetPath()
 
-	// Make sure index isn't out of bound
-	if (len(path.GetPathSegments()) > 0) && int(path.GetIndex()) >= len(path.GetPathSegments()) {
-		return errors.Errorf("NetworkServiceRequest.Connection.Path.Index(%d) >= len(NetworkServiceRequest.Connection.Path.PathSegments)(%d)",
-			path.GetIndex(), len(path.GetPathSegments()))
+	if int(path.Index) < len(path.PathSegments) && path.PathSegments[path.Index].Name == segmentName {
+		// We already in current segment, just update connection Id, no need to increment
+		conn.Id = path.PathSegments[path.Index].Id
+		return conn, nil
 	}
 
-	// If this isn't already one of our tokens for *this* connectionId ... then update the next one
-	if (len(path.GetPathSegments()) > 0) && !(path.GetPathSegments()[path.GetIndex()].GetName() == u.name &&
-		path.GetPathSegments()[path.GetIndex()].GetId() == conn.GetId()) {
-		path.Index++
-	}
+	// We need to move to next item
+	nextIndex := int(path.Index) + 1
 
-	// If we don't already have a PathSegment at the new index... append an empty one
-	if int(path.GetIndex()) >= len(path.GetPathSegments()) {
-		path.PathSegments = append(path.PathSegments, &networkservice.PathSegment{})
+	if nextIndex > len(path.GetPathSegments()) {
+		// We have index > segments count
+		return nil, errors.Errorf("NetworkServiceRequest.Connection.Path.Index+1==%d should be less or equal len(NetworkServiceRequest.Connection.Path.PathSegement)==%d",
+			nextIndex, len(path.GetPathSegments()))
 	}
-
-	// Extract the authInfo:
-	var authInfo credentials.AuthInfo
-	if p, exists := peer.FromContext(ctx); exists {
-		authInfo = p.AuthInfo
-	}
-
-	// Generate the tok
-	tok, expireTime, err := u.tokenGenerator(authInfo)
-	if err != nil {
-		return errors.WithStack(err)
+	if nextIndex < len(path.GetPathSegments()) && path.GetPathSegments()[nextIndex].Name != segmentName {
+		// We have next, but name is different
+		return nil, errors.Errorf("NetworkServiceRequest.Connection.Path.PathSegement[%d].Name should be %s, but it is %s",
```

Comment:
This should not be an error condition... this is a typical healing case, it means that someone in the chain has 'healed' to this server because it couldn't get whomever was previously occupying the path.  If I get a path like through pathSegment names like:

[A,B,C,D,E,F] index 2 and I'm G.  The right move is not to return an error to B, but to substitute myself for C in the path so it becomes [A,B,G,D,E,F]... D may or may not accept the substitution... but that's up to D.

Reply:
Ok, Fixed.